Composer/packages/lib/indexers/src/dialogUtils/dialogChecker.ts

```diff
@@ -11,30 +11,94 @@ import { CheckerFunc } from './types';
 
 const ExpressionParser = new ExpressionEngine();
 
-export const IsExpression: CheckerFunc = (
-  path,
-  value,
-  optional: { properties: string[]; requiredTypes: { [key: string]: boolean } }
-) => {
-  const { properties, requiredTypes } = optional;
-  return properties.reduce((result: Diagnostic[], property) => {
-    let message = '';
-    const exp = get(value, property);
-    if (!exp && requiredTypes[property]) {
-      message = formatMessage(`is missing or empty`);
-    } else {
-      try {
-        ExpressionParser.parse(exp);
-      } catch (error) {
-        message = `${formatMessage('must be an expression:')} ${error})`;
-      }
+const createPath = (path: string, type: string): string => {
+  const steps = ['triggers', 'actions', 'elseActions'];
+  let list = path.split('.');
+  const matchs = list.filter(x => !steps.every(step => !x.startsWith(step)));
+
+  const focused = matchs.join('.');
+  list = path.split(`${focused}.`);
+  if (list.length !== 2) return path;
+
+  return `${list[0]}${focused}#${type}#${list[1]}`;
+};
+
+export const checkExpression = (exp: string, required: boolean, path: string, type: string): Diagnostic | null => {
+  let message = '';
+  if (!exp && required) {
+    message = formatMessage(`is missing or empty`);
+  } else {
+    try {
+      ExpressionParser.parse(exp);
+    } catch (error) {
+      message = `${formatMessage('must be an expression:')} ${error})`;
     }
-    if (message) {
-      const diagnostic = new Diagnostic(message, '');
-      diagnostic.path = `${path}#${value.$type}#${property}`;
-      result.push(diagnostic);
+  }
+  if (message) {
+    const diagnostic = new Diagnostic(message, '');
+    diagnostic.path = createPath(path, type);
+    return diagnostic;
+  }
+
+  return null;
+};
+
+function findAllRequiredType(schema: any): { [key: string]: boolean } {
+  if (!schema) return {};
+  const types = schema.anyOf?.filter(x => x.title === 'Type');
```

Comment:
Where is this `Type` coming from?

Reply:
Hi @cwhitten. I have synced with @luhan2017 and checked the schema. The original schema in [microsoft/botbuilder-dotnet](https://github.com/microsoft/botbuilder-dotnet/tree/master/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Schemas/Actions) use required to set object is required. But in our sdk.schema, use anyof to set.
Microsoft.DeleteProperties:
`
    "required": [
        "properties"
    ],`

`            "anyOf": [
                {
                    "title": "Reference",
                    "required": [
                        "$copy"
                    ]
                },
                {
                    "title": "Type",
                    "required": [
                        "properties",
                        "$kind"
                    ]
                }
            ]`

Microsoft.EditActions:
`
    "required": [
        "changeType",
        "actions"
    ],`

`            "anyOf": [
                {
                    "title": "Reference",
                    "required": [
                        "$copy"
                    ]
                },
                {
                    "title": "Type",
                    "required": [
                        "changeType",
                        "actions",
                        "$kind"
                    ]
                }
            ]`
Microsoft.EditArray:
`
    "required": [
        "changeType",
        "itemsProperty"
    ],`

`            "anyOf": [
                {
                    "title": "Reference",
                    "required": [
                        "$copy"
                    ]
                },
                {
                    "title": "Type",
                    "required": [
                        "changeType",
                        "itemsProperty",
                        "$kind"
                    ]
                }
            ]`

Reply:
So, in short, the schema merging tool will turn the standardized "required" filed in json schema, into a "anyof" field with a hard-coded key "Type", right?